src/cargo/ops/registry.rs

```diff
@@ -90,7 +91,22 @@ pub struct PublishOpts<'cfg> {
 
 pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
     let specs = opts.to_publish.to_package_id_specs(ws)?;
+
+    if let Packages(_) = &opts.to_publish {
+        if specs.len() > 1 {
+            bail!("the `-p` argument must be specified to select a single package to publish");
+        }
+    } else {
+        if ws.is_virtual() {
+            bail!("must use `-p` argument in virtual manifest")
```

Comment:
Just to make the error message wording consistent with the other error.

Also, it would be "with a virtual manifest" or "in a virtual workspace".  "in virtual manifest" would mean inside the `Cargo.toml` file itself.

```suggestion
            bail!("the `-p` argument must be specified in the root of a virtual workspace")
```